tests/test_base_dense.py

```diff
@@ -0,0 +1,49 @@
+from qutip_cupy import CuPyDense
```

Comment:
I would name this file `test_dense.py` to match `dense.py`. In general, most test files should have the name `test_<module-being-tested>.py` although there will be exceptions at times (e.g. maybe we will want to test what is exposed by a package).

Reply:
I would perhaps also organize the tests in the file into classes since there might be a lot of them, e.g.
```python

class TestCuPyDense:
    ... tests of the methods and constructor of CuPyDense go here ...

class TestAdjointSpecialisation:
    ... adjoint tests go here ...

class TestConjSpecialisation:
    ...
```
How exactly to organize things is up for discussion, but it would be good to have some ordering.